src/lib/HST/Frontend/FromGHC.hs

```diff
@@ -0,0 +1,395 @@
+transformExpr (GHC.L s (GHC.ExplicitList _ _ es)) =
+  S.List (transformSrcSpan s) (map transformExpr es)
+transformExpr (GHC.L s (GHC.HsPar _ e)) =
+  S.Paren (transformSrcSpan s) (transformExpr e)
+transformExpr (GHC.L s (GHC.ExprWithTySig _ e typeSig)) =
+  S.ExpTypeSig (transformSrcSpan s) (transformExpr e) (SigType typeSig)
+transformExpr _ = error "Unsupported expression"
+
+-- | Transforms a GHC located tuple argument consisting of an expression into
+--   an HST expression.
+transformTupleArg :: GHC.LHsTupArg GHC.GhcPs -> S.Exp GHC
+transformTupleArg (GHC.L _ (GHC.Present _ e)) = transformExpr e
+transformTupleArg _ = error "Missing argument in tuple"
+
+-- | Transforms a GHC located pattern into an HST pattern.
+transformPat :: GHC.LPat GHC.GhcPs -> S.Pat GHC
+transformPat (GHC.L s (GHC.VarPat _ name)) =
+  S.PVar (transformSrcSpan s) (transformRdrNameUnqual name)
+transformPat (GHC.L s (GHC.ConPatIn name cpds)) =
+  let s'             = transformSrcSpan s
+      (name', isCon) = transformRdrName name
+  in  case (cpds, isCon) of
+        (GHC.InfixCon pat1 pat2, True) ->
+          S.PInfixApp s' (transformPat pat1) name' (transformPat pat2)
+        (GHC.PrefixCon pats, True) -> S.PApp s' name' (map transformPat pats)
+        (_, True) -> error "Record constructors are not supported"
+        _ -> error "Only constructors can be applied in patterns"
+-- TODO The documentation also mentions a more complicated ConPatOut.
+-- Do we need to consider that?
+transformPat (GHC.L s (GHC.TuplePat _ pats boxity)) =
+  S.PTuple (transformSrcSpan s) (transformBoxity boxity) (map transformPat pats)
+transformPat (GHC.L s (GHC.ParPat _ pat)) =
+  S.PParen (transformSrcSpan s) (transformPat pat)
+transformPat (GHC.L s (GHC.ListPat _ pats)) =
+  S.PList (transformSrcSpan s) (map transformPat pats)
+transformPat (GHC.L s (GHC.WildPat _)) = S.PWildCard (transformSrcSpan s)
+transformPat _                         = error "Unsupported pattern"
+
+-- | Transforms a GHC module name with an HST source span into an HST module
+--   name.
+transformModuleName :: S.SrcSpan GHC -> GHC.ModuleName -> S.ModuleName GHC
+transformModuleName s modName = S.ModuleName s (GHC.moduleNameString modName)
+
+-- | Transforms a GHC located reader name into an HST qualified name and a
+--   @Bool@ which is @True@ if the name belongs to a data constructor and
+--   @False@ otherwise.
+transformRdrName :: GHC.Located GHC.RdrName -> (S.QName GHC, Bool)
+transformRdrName (GHC.L s (GHC.Unqual name)) =
+  let s' = transformSrcSpan s
+  in  (S.UnQual s' (S.Ident s' (GHC.occNameString name)), GHC.isDataOcc name)
+transformRdrName (GHC.L s (GHC.Qual modName name)) =
+  let s' = transformSrcSpan s
+  in  ( S.Qual s'
+               (transformModuleName s' modName)
+               (S.Ident s' (GHC.occNameString name))
+      , GHC.isDataOcc name
+      )
+transformRdrName (GHC.L s (GHC.Exact name)) =
+  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), False)
```

Comment:
I was testing with the queue `./example`s and got the following error message whenever the module contained patterns for built-in data types.

> ```
> Internal: Only constructors can be applied in patterns
> CallStack (from HasCallStack):
>   error, called at src/lib/HST/Frontend/FromGHC.hs:324:14 in haskell-src-transformations-0.1.1.0-inplace:HST.Frontend.FromGHC
> ```

As far as I can tell, the problem is that this function returns `False` for special constructors. Do you think that it is safe to change this to `True`?

```suggestion
  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), True)
```

Reply:
Yes, that should be safe. I didn't consider that these built-in types could appear in patterns as well.

Reply:
See a69a8cf230c47c4c20293513a67c42e280f05ffd.